HSearch JSR-352 integration blog post

yrodiere opened the pull request:
Please don't merge yet -- to be published along with the 5.9.0.Beta1 release.

Staging: http://staging.in.relation.to/2017/11/27/introducing-hibernate-search-jsr352-mass-indexing-job/

Sanne commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,158 @@
+:awestruct-layout: blog-post
+---
+
+Hibernate Search 5.9 (still in beta) features a new integration to JSR-352 runtimes,
```
could we reprhase this to also serve as "teaser" for the `5.9.0.Beta1` release?

Proposal:
`Hibernate Search 5.9 will feature integration with JSR-352 [...] Today we released 5.9.0.Beta1 so you can try it out already!`

Sanne commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,158 @@
+
+== What's JSR-352?
+
+"JSR 352: Batch Applications for the Java Platform" is a Java specification
```
To be precise "Java" is no standards body, so either you say "a Java EE specification" or you could say just "a specification" ?

Sanne commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,158 @@
+In JSR-352:
+
+* The job definition is provided as an XML file and a few Java classes, called batch artifacts
+  (in our case, the definition is provided by Hibernate Search).
```
just a style suggestion: try avoid using the parenthesis whenever you can?
In this case the sentence works fine by just removing them.

Sanne commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,158 @@
+* And all that's left is configuring the infrastructure
+  (in our case that's your responsibility, as the Hibernate Search user).
+
+== What's better than the old mass indexer?
```
s/old/existing/ ? :)

or `How is it better than the existing Mass Indexer` ?

Sanne commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,158 @@
+
+== What's better than the old mass indexer?
+
+Ever tried to reindex 100,000 entities with the old mass indexer,
```
s/old/existing/

Sanne commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,158 @@
+== What's better than the old mass indexer?
+
+Ever tried to reindex 100,000 entities with the old mass indexer,
+only to experience some network or disk error at the 99,001th entity?
```
hey that never happened! :D

Sanne commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,158 @@
+probably somewhere around the 95,000th entity.
+
+There are other advantages (such as being able to pool resources between multiple, unrelated background jobs),
+but this one really is why we think JSR-352 makes sense.
```
`this one really is why we think` doesn't flow very well in English. "why" isn't the same as "the reason for" - as in latin languages..

yrodiere replied:
Ok, I'll change that, but... I frequently see "this is why + subject + verb" or similar phrasing... Is there something special here that makes it incorrect?

Sanne replied:
it's indeed slightly different. The "why" is to answer the question, so "this is why xyz" is similar to "this is to answer the question you might have about xyz" but you can't say "xyz is the why"

Sanne commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,158 @@
+
+=== Setup
+
+Within Wildfly https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-configuration-deploy-on-wildfly[use our WildFly modules],
```
Let's not call them `WildFly modules` anymore ;)

Sanne commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
````diff
@@ -0,0 +1,158 @@
+To resume the job after suspension or upon failure, you can request a restart this way:
+
+```
+long executionId2 = jobOperator.restart( executionId, parameters );
````
Is there a way to force a full restart? Use case: it's running, then I realize it's not using the analyzer configurations I want, to I kill it with the intent to have it start over.

yrodiere replied:
You just start another execution (`.start`) with the same parameters. But I'm not sure it would be useful to mention it here... ?

Sanne replied:
sure, np. Was just a thing which came to mind, wanted to make sure that's possible.

Sanne commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,158 @@
+== Credits
+
+I would like to mention that a large part of this JSR-352 integration was developed
+by https://github.com/mincong-h[Mincong Huang] during and after his Google Summer of Code.
```
s/Google Summer of Code/Google Summer of Code project/ ?

Sanne commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,158 @@
+https://hibernate.atlassian.net/browse/HSEARCH-2956[with other JEE application servers],
+and hopefully https://hibernate.atlassian.net/browse/HSEARCH-2692[with Spring's own Spring Batch project],
+which has limited support for JSR-352.
+
```
Could you insert a "try it out now!" paragraph reminding the `5.9.0.Beta1` release and MAven Gav ?

yrodiere commented:
@Sanne I amended my commit to address your comments. Thanks!

Sanne commented:
awesome, that was fast!

You agree we'll publish this monday or maybe thuesday? Being thanksgiving in US and quite late in EU I'd rather wait a little than rush it today.

gunnarmorling commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,163 @@
+:awestruct-layout: blog-post
+---
+
+Hibernate Search 5.9 will feature integration with JSR-352 runtimes,
```
s/JSR-352/JSR 352/.

Also it should be shortly spelled out at the first occurence for people who don't know what it is (I know it's coming below, but I think should be done right here).

gunnarmorling commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,163 @@
+
+== What's JSR-352?
+
+"JSR 352: Batch Applications for the Java Platform" is a Java EE specification
```
It's not targeting Jave EE exclusively.

Sanne replied:
damn you're essentially suggesting the opposite of what I suggested before :D 
 - https://github.com/hibernate/in.relation.to/pull/98#discussion_r153006248

he's not saying that it's exclusive to Java EE, yet it is a Java EE specification.

Maybe just "specification" to make everyone happy? or ignore us Yoann ;)

gunnarmorling replied:
Just "specification" is good. I'd equally cringe if you said BV is a Java EE specification. Yes, it is part of EE, but it also runs under SE. Saying it's a EE spec obfuscates this fact.

gunnarmorling commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,163 @@
+  In our case, the definition is provided by Hibernate Search.
+* The JSR-352 runtime avoids boilerplate code by managing
+  threading, transactions, the execution flow (calling methods on batch artifacts), ...
+* And all that's left is configuring the infrastructure
```
Dot missing at the end.

gunnarmorling commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,163 @@
+=== Setup
+
+Within Wildfly https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-configuration-deploy-on-wildfly[use our JBoss modules],
+configure Hibernate ORM and Hibernate Search. Also add a dependency to the `org.hibernate.search.jsr352` module.
```
How do I add that dependency?

yrodiere replied:
The same you would do with any other dependency to a WildFly. I'm not explaining how to add a dependency to the POM, I don't see why I would explain how to add a dependency to a WildFly deployment...

I'll add "to your deployment" in this sentence, maybe that's what you meant.

gunnarmorling replied:
Adding dependencies to a POM is common knowledge, whereas how to add specific module dependencies to a WF deployment is known to 1% of people. Also does it need the export flag for instance?

> In order for your application to access the Hibernate JSR 352 batch, [add a dependency](https://docs.jboss.org/author/display/WFLY10/Class+Loading+in+WildFly#ClassLoadinginWildFly-JBossDeploymentStructureFile) to the `org.hibernate.search.jsr352` module to your deployment, e.g. by specifying the following in your application's _META-INF/MANIFEST.MF_ file:
> `Dependencies: org.hibernate.search.jsr352`

That way all the info needed is there and people interested have a reference to more details.

gunnarmorling commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,163 @@
+That's all: you can then retrieve a `javax.batch.operations.JobOperator` instance
+in your application through CDI injection.
+
+In a Java SE application, you can use the reference implementation for JSR-352,
```
How about recommending our own one instead, JBeret?

yrodiere replied:
It's really awful in Java SE. The JBeret guys defined all of their dependencies as provided in their POM, so you basically have to add each and every transitive dependency explicitly to your POM, with the correct version. A PITA, really. And they seem to be aware of this, since their documentation includes a chapter instructing users to add these dependencies in their POM.

Besides, mentioning something else than JBeret shows that we're not targeting a specific runtime.

yrodiere replied:
I will add a link to the JBeret documentation: at least the documentation is good. People will be able to make their own opinion.

gunnarmorling commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,163 @@
+** `com.ibm.jbatch:com.ibm.jbatch-runtime:1.0`
+** `javax.batch:javax.batch-api:1.0`
+** `javax.inject:javax.inject:1`
+** `org.apache.derby:derby:10.13.1.1`
```
Derby or any other database?

yrodiere replied:
Probably others too, but JBatch uses hardcoded SQL statements, without a dialect layer, so not just any database will do. H2 doesn't seem to work.
Anyway... This should get people started. If they want more, they can have a look at the configuration file, or switch to another JSR 352 provider (that's the point of having a spec, isn't it?).

gunnarmorling commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,163 @@
+  and to only shut it down after the last mass indexing job stopped.
+
+As for other JSR-352 runtimes, the mass indexer may or may not work.
+We are in the process of testing at least one other JEE application server and Spring Batch:
```
s/JEE/Java EE/

gunnarmorling commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,163 @@
+* `.purgeAllOnStart(boolean)`: specify whether the existing index should be purged at the beginning of the job.
+* `.restrictedBy(Criterion)`: use a `Criterion` to only index a subset of entities.
+* `.restrictedBy(String)`: use HQL to only index a subset of entities.
+  Note that when using HQL, the job switches to a degraded mode that suffers from several limitations,
```
That sounds mysterious, what is it about?

yrodiere replied:
Mainly no partitioning (because we don't use a list of IDs to drive the indexing, but the result of a query returning entities). So HQL indexing will use a single thread. We might try to change that, rely on limit/offset to run the query in multiple threads, but then there will be a risk of indexing some entities twice or skipping some entities. Either way, not a good idea to use HQL when you could use something else.

gunnarmorling commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,163 @@
+
+== Credits
+
+I would like to mention that a large part of this JSR-352 integration was developed
```
I think this should be mentioned right at the beginning (E.g. "Originally started as a GSoC project by Mincong Huang, we're today happy to announce the release of ...").

gunnarmorling commented on posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc:
```diff
@@ -0,0 +1,163 @@
+with two JSR-352 runtimes: JBatch in standalone mode and JBeret in a WildFly server.
+
+We would like to make sure our mass indexing job will work
+https://hibernate.atlassian.net/browse/HSEARCH-2956[with other JEE application servers],
```
s/JEE/Java EE/

yrodiere commented:
> You agree we'll publish this monday or maybe thuesday? Being thanksgiving in US and quite late in EU I'd rather wait a little than rush it today.

A bit late, but yes I agree. I wasn't even considering releasing on a Friday evening ;)

@gunnarmorling I addressed your comments, thanks!
